[R-Package] CRAN issues

## Environment info
Operating System: Windows 8.1 Pro
CPU: i7-4600U
R version: 3.4

To make a release on CRAN, we will need first to fix all the errors / warnings / notes. Currently testing on Windows, but we will also need to test on Linux. If some of them cannot be fixed, we will need to have an explanation for each of those which will not be fixed by us. @guolinke 

Maybe time to add some vignettes @coforfe if you want to work on them.

- [x] Fix CRAN errors
- [x] Fix CRAN warnings
- [x] Fix CRAN notes
- [x] Add vignettes
- [x] Make examples runnable for tests
- [x] Switch to MinGW by default on Windows (it falls to MinGW if VS fails to be found)
- [x] Clean install everytime when not using precompiled dll/lib
- [x] Pass CRAN checks on Windows
- [x] Pass CRAN checks on Linux
- [x] Make CRAN-valid portable installation configuration (#2960)
- [x] Submit to CRAN
- [x] Get accepted on CRAN

00install.out:

```r
* installing *source* package 'lightgbm' ...
** libs
installing via 'install.libs.R' to C:/tmp/RtmpG2NQSv/lightgbm.Rcheck/lightgbm
Warning in file.copy("./../../include", "./", overwrite = TRUE, recursive = TRUE) :
  problem copying .\..\..\include to .\include: No such file or directory
Error in eval(ei, envir) : Cannot find folder LightGBM/include
* removing 'C:/tmp/RtmpG2NQSv/lightgbm.Rcheck/lightgbm'
```

Windows (fake) CRAN log:

```r
R version 3.4.0 (2017-04-21) -- "You Stupid Darkness"
Copyright (C) 2017 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library(devtools)
> library(roxygen2) # devtools::install_github("klutometis/roxygen")
> setwd("C:/LightGBM/R-package")
> devtools::check()
Updating lightgbm documentation
Loading lightgbm
Re-compiling lightgbm
"C:/PROGRA~1/R/R-34~1.0/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD INSTALL "C:\LightGBM\R-package"  \
  --library="C:\tmp\RtmpG2NQSv\devtools_install_19fa019043416" --no-R --no-data --no-help --no-demo --no-inst --no-docs --no-exec --no-multiarch  \
  --no-test-load --preclean 

* installing *source* package 'lightgbm' ...
** libs
installing via 'install.libs.R' to C:/tmp/RtmpG2NQSv/devtools_install_19fa019043416/lightgbm
-- Building for: Visual Studio 15 2017
-- The C compiler identification is MSVC 19.11.25303.0
-- The CXX compiler identification is MSVC 19.11.25303.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/Preview/Community/VC/Tools/MSVC/14.11.25301/bin/HostX86/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/Preview/Community/VC/Tools/MSVC/14.11.25301/bin/HostX86/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/Preview/Community/VC/Tools/MSVC/14.11.25301/bin/HostX86/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/Preview/Community/VC/Tools/MSVC/14.11.25301/bin/HostX86/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Try OpenMP C flag = [/openmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Success
-- Try OpenMP CXX flag = [/openmp]
-- Performing Test OpenMP_FLAG_DETECTED
-- Performing Test OpenMP_FLAG_DETECTED - Success
-- Found OpenMP: /openmp  
-- Configuring done
-- Generating done
-- Build files have been written to: C:/LightGBM/R-package/src/build
Microsoft (R) Build Engine version 15.3.118.39484
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 16/06/2017 07:00:42 PM.
Project "C:\LightGBM\R-package\src\build\_lightgbm.vcxproj" on node 1 (default targets).
Project "C:\LightGBM\R-package\src\build\_lightgbm.vcxproj" (1) is building "C:\LightGBM\R-package\src\build\ZERO_CHECK.vcxproj" (2) on node 1 (default targets).
PrepareForBuild:
  Creating directory "x64\Release\ZERO_CHECK\".
  Creating directory "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\".
InitializeBuildStatus:
  Creating "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Checking Build System
  CMake does not need to re-run because C:/LightGBM/R-package/src/build/CMakeFiles/generate.stamp is up-to-date.
FinalizeBuildStatus:
  Deleting file "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
  Touching "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
Done Building Project "C:\LightGBM\R-package\src\build\ZERO_CHECK.vcxproj" (default targets).
PrepareForBuild:
  Creating directory "_lightgbm.dir\Release\".
  Creating directory "C:\LightGBM\R-package\src\Release\".
  Creating directory "_lightgbm.dir\Release\_lightgbm.tlog\".
InitializeBuildStatus:
  Creating "_lightgbm.dir\Release\_lightgbm.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Building Custom Rule C:/LightGBM/R-package/src/CMakeLists.txt
  CMake does not need to re-run because C:/LightGBM/R-package/src/build/CMakeFiles/generate.stamp is up-to-date.
ClCompile:
  C:\Program Files (x86)\Microsoft Visual Studio\Preview\Community\VC\Tools\MSVC\14.11.25301\bin\HostX86\x64\CL.exe /c /I"C:\LightGBM\R-package\src\include" /nologo /W4 /WX- /diagnostics:classic /O2 /Ob2 /Oi /Ot /Oy /GL /D WIN32 /D _WINDOWS /D NDEBUG /D USE_SOCKET /D "CMAKE_INTDIR=\"Release\"" /D _lightgbm_EXPORTS /D _WINDLL /D _MBCS /UNDEBUG /Gm- /EHsc /MD /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /openmp /Fo"_lightgbm.dir\Release\\" /Fd"_lightgbm.dir\Release\vc141.pdb" /Gd /TP /errorReport:queue   -pedantic -g -O0 "C:\LightGBM\R-package\src\src\c_api.cpp" "C:\LightGBM\R-package\src\src\lightgbm_R.cpp" "C:\LightGBM\R-package\src\src\application\application.cpp" "C:\LightGBM\R-package\src\src\boosting\boosting.cpp" "C:\LightGBM\R-package\src\src\boosting\gbdt.cpp" "C:\LightGBM\R-package\src\src\boosting\gbdt_prediction.cpp" "C:\LightGBM\R-package\src\src\boosting\prediction_early_stop.cpp" "C:\LightGBM\R-package\src\src\io\bin.cpp" "C:\LightGBM\R-package\src\src\io\config.cpp" "C:\LightGBM\R-package\src\src\io\dataset.cpp" "C:\LightGBM\R-package\src\src\io\dataset_loader.cpp" "C:\LightGBM\R-package\src\src\io\metadata.cpp" "C:\LightGBM\R-package\src\src\io\parser.cpp" "C:\LightGBM\R-package\src\src\io\tree.cpp" "C:\LightGBM\R-package\src\src\metric\dcg_calculator.cpp" "C:\LightGBM\R-package\src\src\metric\metric.cpp" "C:\LightGBM\R-package\src\src\objective\objective_function.cpp" "C:\LightGBM\R-package\src\src\network\linker_topo.cpp" "C:\LightGBM\R-package\src\src\network\linkers_mpi.cpp" "C:\LightGBM\R-package\src\src\network\linkers_socket.cpp" "C:\LightGBM\R-package\src\src\network\network.cpp" "C:\LightGBM\R-package\src\src\treelearner\data_parallel_tree_learner.cpp" "C:\LightGBM\R-package\src\src\treelearner\feature_parallel_tree_learner.cpp" "C:\LightGBM\R-package\src\src\treelearner\gpu_tree_learner.cpp" "C:\LightGBM\R-package\src\src\treelearner\serial_tree_learner.cpp" "C:\LightGBM\R-package\src\src\treelearner\tree_learner.cpp" "C:\LightGBM\R-package\src\src\treelearner\voting_parallel_tree_learner.cpp"
cl : Command line warning D9025: overriding '/DNDEBUG' with '/UNDEBUG' [C:\LightGBM\R-package\src\build\_lightgbm.vcxproj]
  c_api.cpp
cl : Command line warning D9002: ignoring unknown option '-pedantic' [C:\LightGBM\R-package\src\build\_lightgbm.vcxproj]
cl : Command line warning D9002: ignoring unknown option '-g' [C:\LightGBM\R-package\src\build\_lightgbm.vcxproj]
cl : Command line warning D9002: ignoring unknown option '-O0' [C:\LightGBM\R-package\src\build\_lightgbm.vcxproj]
  lightgbm_R.cpp
  application.cpp
  boosting.cpp
  gbdt.cpp
  gbdt_prediction.cpp
  prediction_early_stop.cpp
  bin.cpp
  config.cpp
  dataset.cpp
  dataset_loader.cpp
  metadata.cpp
  parser.cpp
  tree.cpp
  dcg_calculator.cpp
  metric.cpp
  objective_function.cpp
  linker_topo.cpp
  linkers_mpi.cpp
  linkers_socket.cpp
  Compiling...
  network.cpp
  data_parallel_tree_learner.cpp
  feature_parallel_tree_learner.cpp
  gpu_tree_learner.cpp
  serial_tree_learner.cpp
  tree_learner.cpp
  voting_parallel_tree_learner.cpp
Link:
  C:\Program Files (x86)\Microsoft Visual Studio\Preview\Community\VC\Tools\MSVC\14.11.25301\bin\HostX86\x64\link.exe /ERRORREPORT:QUEUE /OUT:"C:\LightGBM\R-package\src\Release\lib_lightgbm.dll" /INCREMENTAL:NO /NOLOGO kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /PDB:"C:/LightGBM/R-package/src/Release/lib_lightgbm.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"C:/LightGBM/R-package/src/Release/lib_lightgbm.lib" /MACHINE:X64   /machine:x64 /DLL _lightgbm.dir\Release\c_api.obj
  _lightgbm.dir\Release\lightgbm_R.obj
  _lightgbm.dir\Release\application.obj
  _lightgbm.dir\Release\boosting.obj
  _lightgbm.dir\Release\gbdt.obj
  _lightgbm.dir\Release\gbdt_prediction.obj
  _lightgbm.dir\Release\prediction_early_stop.obj
  _lightgbm.dir\Release\bin.obj
  _lightgbm.dir\Release\config.obj
  _lightgbm.dir\Release\dataset.obj
  _lightgbm.dir\Release\dataset_loader.obj
  _lightgbm.dir\Release\metadata.obj
  _lightgbm.dir\Release\parser.obj
  _lightgbm.dir\Release\tree.obj
  _lightgbm.dir\Release\dcg_calculator.obj
  _lightgbm.dir\Release\metric.obj
  _lightgbm.dir\Release\objective_function.obj
  _lightgbm.dir\Release\linker_topo.obj
  _lightgbm.dir\Release\linkers_mpi.obj
  _lightgbm.dir\Release\linkers_socket.obj
  _lightgbm.dir\Release\network.obj
  _lightgbm.dir\Release\data_parallel_tree_learner.obj
  _lightgbm.dir\Release\feature_parallel_tree_learner.obj
  _lightgbm.dir\Release\gpu_tree_learner.obj
  _lightgbm.dir\Release\serial_tree_learner.obj
  _lightgbm.dir\Release\tree_learner.obj
  _lightgbm.dir\Release\voting_parallel_tree_learner.obj
  c_api.obj : MSIL .netmodule or module compiled with /GL found; restarting link with /LTCG; add /LTCG to the link command line to improve linker performance
     Creating library C:/LightGBM/R-package/src/Release/lib_lightgbm.lib and object C:/LightGBM/R-package/src/Release/lib_lightgbm.exp
  Generating code
  Finished generating code
  _lightgbm.vcxproj -> C:\LightGBM\R-package\src\Release\lib_lightgbm.dll
FinalizeBuildStatus:
  Deleting file "_lightgbm.dir\Release\_lightgbm.tlog\unsuccessfulbuild".
  Touching "_lightgbm.dir\Release\_lightgbm.tlog\_lightgbm.lastbuildstate".
Done Building Project "C:\LightGBM\R-package\src\build\_lightgbm.vcxproj" (default targets).

Build succeeded.

"C:\LightGBM\R-package\src\build\_lightgbm.vcxproj" (default target) (1) ->
(ClCompile target) -> 
  cl : Command line warning D9025: overriding '/DNDEBUG' with '/UNDEBUG' [C:\LightGBM\R-package\src\build\_lightgbm.vcxproj]
  cl : Command line warning D9002: ignoring unknown option '-pedantic' [C:\LightGBM\R-package\src\build\_lightgbm.vcxproj]
  cl : Command line warning D9002: ignoring unknown option '-g' [C:\LightGBM\R-package\src\build\_lightgbm.vcxproj]
  cl : Command line warning D9002: ignoring unknown option '-O0' [C:\LightGBM\R-package\src\build\_lightgbm.vcxproj]

    4 Warning(s)
    0 Error(s)

Time Elapsed 00:00:44.73
Found library file: C:/LightGBM/R-package/src/Release/lib_lightgbm.dll to move to C:/tmp/RtmpG2NQSv/devtools_install_19fa019043416/lightgbm/libs/x64* DONE (lightgbm)
Loading required package: R6
Writing NAMESPACE
Writing predict.lgb.Booster.Rd
Setting env vars -------------------------------------------------------------------------------------------------------------------------------------------
CFLAGS  : -Wall -pedantic
CXXFLAGS: -Wall -pedantic
Building lightgbm ------------------------------------------------------------------------------------------------------------------------------------------
"C:/PROGRA~1/R/R-34~1.0/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD build "C:\LightGBM\R-package" --no-resave-data  \
  --no-manual 

* checking for file 'C:\LightGBM\R-package/DESCRIPTION' ... OK
* preparing 'lightgbm':
* checking DESCRIPTION meta-information ... OK
* cleaning src
* checking for LF line-endings in source and make files
* checking for empty or unneeded directories
Removed empty directory 'lightgbm/src/build/CMakeFiles/3.8.0/CompilerIdC/tmp'
Removed empty directory 'lightgbm/src/build/CMakeFiles/3.8.0/CompilerIdCXX/tmp'
Removed empty directory 'lightgbm/src/build/CMakeFiles/CMakeTmp'
* looking to see if a 'data/datalist' file should be added
* building 'lightgbm_0.2.tar.gz'
Warning in utils::tar(filepath, pkgname, compression = "gzip", compression_level = 9L,  :
  storing paths of more than 100 bytes is not portable:
  'lightgbm/src/build/CMakeFiles/3.8.0/CompilerIdCXX/Debug/CompilerIdCXX.tlog/CompilerIdCXX.lastbuildstate'

Setting env vars -------------------------------------------------------------------------------------------------------------------------------------------
_R_CHECK_CRAN_INCOMING_ : FALSE
_R_CHECK_FORCE_SUGGESTS_: FALSE
Checking lightgbm ------------------------------------------------------------------------------------------------------------------------------------------
"C:/PROGRA~1/R/R-34~1.0/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD check "C:\tmp\RtmpG2NQSv/lightgbm_0.2.tar.gz" --as-cran  \
  --timings --no-manual 

* using log directory 'C:/tmp/RtmpG2NQSv/lightgbm.Rcheck'
* using R version 3.4.0 (2017-04-21)
* using platform: x86_64-w64-mingw32 (64-bit)
* using session charset: ISO8859-1
* using options '--no-manual --as-cran'
* checking for file 'lightgbm/DESCRIPTION' ... OK
* checking extension type ... Package
* this is package 'lightgbm' version '0.2'
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... WARNING
Subdirectory 'src' contains:
  CMakeLists.txt
These are unlikely file names for src files.

In addition to the above warning(s), found the following notes:

Found the following apparent object files/libraries:
  src/Release/lib_lightgbm.dll
  src/build/CMakeFiles/3.8.0/CompilerIdC/Debug/CMakeCCompilerId.obj
  src/build/CMakeFiles/3.8.0/CompilerIdCXX/Debug/CMakeCXXCompilerId.obj
  src/build/_lightgbm.dir/Release/application.obj
  src/build/_lightgbm.dir/Release/bin.obj
  src/build/_lightgbm.dir/Release/boosting.obj
  src/build/_lightgbm.dir/Release/c_api.obj
  src/build/_lightgbm.dir/Release/config.obj
  src/build/_lightgbm.dir/Release/data_parallel_tree_learner.obj
  src/build/_lightgbm.dir/Release/dataset.obj
  src/build/_lightgbm.dir/Release/dataset_loader.obj
  src/build/_lightgbm.dir/Release/dcg_calculator.obj
  src/build/_lightgbm.dir/Release/feature_parallel_tree_learner.obj
  src/build/_lightgbm.dir/Release/gbdt.obj
  src/build/_lightgbm.dir/Release/gbdt_prediction.obj
  src/build/_lightgbm.dir/Release/gpu_tree_learner.obj
  src/build/_lightgbm.dir/Release/lightgbm_R.obj
  src/build/_lightgbm.dir/Release/linker_topo.obj
  src/build/_lightgbm.dir/Release/linkers_mpi.obj
  src/build/_lightgbm.dir/Release/linkers_socket.obj
  src/build/_lightgbm.dir/Release/metadata.obj
  src/build/_lightgbm.dir/Release/metric.obj
  src/build/_lightgbm.dir/Release/network.obj
  src/build/_lightgbm.dir/Release/objective_function.obj
  src/build/_lightgbm.dir/Release/parser.obj
  src/build/_lightgbm.dir/Release/prediction_early_stop.obj
  src/build/_lightgbm.dir/Release/serial_tree_learner.obj
  src/build/_lightgbm.dir/Release/tree.obj
  src/build/_lightgbm.dir/Release/tree_learner.obj
  src/build/_lightgbm.dir/Release/voting_parallel_tree_learner.obj
Object files/libraries should not be included in a source package.
* checking if there is a namespace ... OK
* checking for executable files ... WARNING
Found the following executable files:
  src/Release/lib_lightgbm.dll
  src/build/CMakeFiles/3.8.0/CMakeDetermineCompilerABI_C.bin
  src/build/CMakeFiles/3.8.0/CMakeDetermineCompilerABI_CXX.bin
  src/build/CMakeFiles/3.8.0/CompilerIdC/CompilerIdC.exe
  src/build/CMakeFiles/3.8.0/CompilerIdCXX/CompilerIdCXX.exe
  src/build/CMakeFiles/FindOpenMP/ompver_C.bin
  src/build/CMakeFiles/FindOpenMP/ompver_CXX.bin
  src/build/CMakeFiles/feature_tests.bin
Source packages should not contain undeclared executable files.
See section 'Package structure' in the 'Writing R Extensions' manual.
* checking for hidden files and directories ... OK
* checking for portable file names ... NOTE
Found the following non-portable file path:
  lightgbm/src/build/CMakeFiles/3.8.0/CompilerIdCXX/Debug/CompilerIdCXX.tlog/CompilerIdCXX.lastbuildstate

Tarballs are only required to store paths of up to 100 bytes and cannot
store those of more than 256 bytes, with restrictions including to 100
bytes for the final component.
See section 'Package structure' in the 'Writing R Extensions' manual.
* checking whether package 'lightgbm' can be installed ...Warning: running command '"C:/PROGRA~1/R/R-34~1.0/bin/x64/Rcmd.exe" INSTALL -l "C:/tmp/RtmpG2NQSv/lightgbm.Rcheck" --no-html --no-multiarch "C:\tmp\RTMPG2~1\LIGHTG~1.RCH\00_PKG~1\lightgbm"' had status 1
 ERROR
Installation failed.
See 'C:/tmp/RtmpG2NQSv/lightgbm.Rcheck/00install.out' for details.
* DONE

Status: 1 ERROR, 2 WARNINGs, 1 NOTE
See
  'C:/tmp/RtmpG2NQSv/lightgbm.Rcheck/00check.log'
for details.

R CMD check results
1 error  | 2 warnings | 1 note 
checking whether package 'lightgbm' can be installed ... ERROR
Installation failed.
See 'C:/tmp/RtmpG2NQSv/lightgbm.Rcheck/00install.out' for details.

checking if this is a source package ... WARNING
Subdirectory 'src' contains:
  CMakeLists.txt
These are unlikely file names for src files.

In addition to the above warning(s), found the following notes:

Found the following apparent object files/libraries:
  src/Release/lib_lightgbm.dll
  src/build/CMakeFiles/3.8.0/CompilerIdC/Debug/CMakeCCompilerId.obj
... 19 lines ...
  src/build/_lightgbm.dir/Release/metadata.obj
  src/build/_lightgbm.dir/Release/metric.obj
  src/build/_lightgbm.dir/Release/network.obj
  src/build/_lightgbm.dir/Release/objective_function.obj
  src/build/_lightgbm.dir/Release/parser.obj
  src/build/_lightgbm.dir/Release/prediction_early_stop.obj
  src/build/_lightgbm.dir/Release/serial_tree_learner.obj
  src/build/_lightgbm.dir/Release/tree.obj
  src/build/_lightgbm.dir/Release/tree_learner.obj
  src/build/_lightgbm.dir/Release/voting_parallel_tree_learner.obj
Object files/libraries should not be included in a source package.

checking for executable files ... WARNING
Found the following executable files:
  src/Release/lib_lightgbm.dll
  src/build/CMakeFiles/3.8.0/CMakeDetermineCompilerABI_C.bin
  src/build/CMakeFiles/3.8.0/CMakeDetermineCompilerABI_CXX.bin
  src/build/CMakeFiles/3.8.0/CompilerIdC/CompilerIdC.exe
  src/build/CMakeFiles/3.8.0/CompilerIdCXX/CompilerIdCXX.exe
  src/build/CMakeFiles/FindOpenMP/ompver_C.bin
  src/build/CMakeFiles/FindOpenMP/ompver_CXX.bin
  src/build/CMakeFiles/feature_tests.bin
Source packages should not contain undeclared executable files.
See section 'Package structure' in the 'Writing R Extensions' manual.

checking for portable file names ... NOTE
Found the following non-portable file path:
  lightgbm/src/build/CMakeFiles/3.8.0/CompilerIdCXX/Debug/CompilerIdCXX.tlog/CompilerIdCXX.lastbuildstate

Tarballs are only required to store paths of up to 100 bytes and cannot
store those of more than 256 bytes, with restrictions including to 100
bytes for the final component.
See section 'Package structure' in the 'Writing R Extensions' manual.
```

Ok I took the current version of the R package on #3188 , since I think that PR will be merged  soon, and built an R source package to check on R Hub.

<details><summary>How to run R Hub checks</summary>

from a terminal:

```shell
sh build-cran-package.sh
```

from R:

```r
result <- rhub::check(
    path = "lightgbm_2.3.2.tar.gz"
    , email = "jaylamb20@gmail.com"
    , check_args = "--as-cran"
    , platform = rhub::platforms()[["name"]]
    , env_vars = c(
        "R_COMPILE_AND_INSTALL_PACKAGES" = "always"
    )
)
```

</details>

## Overview

[R Hub](https://builder.r-hub.io/about.html) is a free project that runs `R CMD check` on an R package on a bunch of platforms.

It doesn't respond fast enough to be part of our normal CI, and it is independent of CRAN so the environments there are slightly different, but it is a free way to test on a lot of setups, especially older Windows environments.

The results are below. I don't know how long the log links last, but I included those as well

## Results

❌  **`Oracle Solaris 10, x86, 32 bit, R-release, Oracle Developer Studio 12.6`**

<details><summary>1 NOTE, 0 WARNINGS, 1 ERROR</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-27612bec0da64d65a24cdcc48c376ded)

```text
/opt/csw/bin/g++-5.5 -std=gnu++11 -I"/opt/csw/lib/R/include" -DNDEBUG -I./include -DMM_PREFETCH=1 -DMM_MALLOC=1 -DUSE_SOCKET -DLGB_R_BUILD  -I/opt/csw/include  -pthread -fPIC  -O2 -pipe -fno-omit-frame-pointer -m32 -march=pentiumpro  -c network/linkers_socket.cpp -o network/linkers_socket.o
In file included from network/linkers.h:22:0,
                 from network/linkers_socket.cpp:19:
network/socket_wrapper.hpp:30:21: fatal error: ifaddrs.h: No such file or directory
compilation terminated.
gmake: *** [/opt/csw/lib/R/etc/Makeconf:176: network/linkers_socket.o] Error 1
ERROR: compilation failed for package ‘lightgbm’
* removing ‘/export/home/X8LyD0f/R/lightgbm’
Warning message:
In i.p(...) :
  installation of package ‘/export/home/X8LyD0f/Rtemp/RtmpMrj4mJ/file618843243bf5/lightgbm_2.3.2.tar.gz’ had non-zero exit status
```

</details>


❌  **`Oracle Solaris 10, x86, 32 bit, R-release`**

<details><summary>1 NOTE, 0 WARNINGS, 1 ERROR</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-de74e238a06d4265a5764f5474b34f29)

```text
/opt/csw/bin/g++-5.5 -std=gnu++11 -I"/opt/csw/lib/R/include" -DNDEBUG -I./include -DMM_PREFETCH=1 -DMM_MALLOC=1 -DUSE_SOCKET -DLGB_R_BUILD  -I/opt/csw/include  -pthread -fPIC  -O2 -pipe -fno-omit-frame-pointer -m32 -march=pentiumpro  -c network/linkers_socket.cpp -o network/linkers_socket.o
In file included from network/linkers.h:22:0,
                 from network/linkers_socket.cpp:19:
network/socket_wrapper.hpp:30:21: fatal error: ifaddrs.h: No such file or directory
compilation terminated.
gmake: *** [/opt/csw/lib/R/etc/Makeconf:176: network/linkers_socket.o] Error 1
ERROR: compilation failed for package ‘lightgbm’
```

</details>

✅ **`macOS 10.13.6 High Sierra, R-release, CRAN's setup`**

<details><summary>1 NOTE, 2 WARNINGS, 0 ERRORs</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-39a81f9ed7c348a8bb1e6081a103ef86)

```text
* checking whether package ‘lightgbm’ can be installed ... WARNING
Found the following significant warnings:
  Warning: package ‘R6’ was built under R version 4.0.2
See ‘/Users/userllqNxUln/lightgbm.Rcheck/00install.out’ for details.

* checking top-level files ... WARNING
A complete check needs the 'checkbashisms' script.
See section ‘Configure and cleanup’ in the ‘Writing R Extensions’
manual.
```

</details>

✅ **`macOS 10.13.6 High Sierra, R-release, brew`**

<details><summary>1 NOTE, 2 WARNINGS, 0 ERRORs</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-a119459a3b62462a9ba62c2435aff183)

```text
* checking whether package ‘lightgbm’ can be installed ... WARNING
Found the following significant warnings:
  Warning: package ‘R6’ was built under R version 4.0.2
See ‘/Users/userllqNxUln/lightgbm.Rcheck/00install.out’ for details.

* checking top-level files ... WARNING
A complete check needs the 'checkbashisms' script.
See section ‘Configure and cleanup’ in the ‘Writing R Extensions’
manual.
```

</details>

⚠️ **`Ubuntu Linux 16.04 LTS, R-release, GCC`**

<details><summary>3 NOTEs, 0 WARNINGS, 0 ERRORs</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-378a2141f81d4a82bc64f4334d6e3fd7)

```text
* checking installed package size ... NOTE
  installed size is 35.7Mb
  sub-directories of 1Mb or more:
    libs  35.1Mb

* checking examples ... NOTE
Examples with CPU or elapsed time > 5s
                      user system elapsed
lgb.cv              22.910  0.197  70.190
lgb.model.dt.tree   13.948  0.141  41.967
lgb.importance       6.955  0.058  20.697
lgb.plot.importance  6.864  0.083  19.118
lgb.interprete       4.675  0.074  14.342
lgb.get.eval.result  4.336  0.049  12.555
lgb.unloader         4.261  0.005  12.519
lgb.train            4.200  0.025  13.718
predict.lgb.Booster  4.159  0.003  12.443
```

</details>


⚠️ **`Fedora Linux, R-devel, clang, gfortran`**

<details><summary>build timed out after 20 mins</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-c9630359a10845f9810ef2a6df8130dd)

</details>

✅ **`Windows Server 2008 R2 SP1, R-oldrel, 32/64 bit`**

<details><summary>1 NOTE, 0 WARNINGS, 0 ERRORs</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-28271b3d8f3347a7bd647b25b68908d8)

</details>

✅ **`Windows Server 2008 R2 SP1, R-release, 32/64 bit`**

<details><summary>1 NOTE, 0 WARNINGs, 0 ERRORs</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-0c2d3bbb8fa74becacc683b8c1b27268)

</details>

✅ **`Windows Server 2008 R2 SP1, R-patched, 32/64 bit`**

<details><summary>NOTEs, WARNINGS, ERRORs</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-5e99751523234b398aad73250062de32)

</details>

✅ **`Windows Server 2008 R2 SP1, R-devel, 32/64 bit`**

<details><summary>NOTEs, WARNINGS, ERRORs</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-8ceed10810ca4fc88d7fa95b7f14f2d8)

</details>

✅ **`Windows Server 2008 R2 SP1, R-release, 32/64 bit`**

<details><summary>1 NOTE, 0 WARNINGs, 0 ERRORs</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-dbd2092382ea4478866cfcef4c4ef5cf)

</details>

⚠️ **`Ubuntu Linux 16.04 LTS, R-devel, GCC`**

<details><summary>3 NOTEs, 1 WARNING, 0 ERRORs</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-6bb8a4e52c4a4b6cbfc4da79922619ac)

```text
* checking installed package size ... NOTE
  installed size is 11.6Mb
  sub-directories of 1Mb or more:
    libs  11.0Mb

* checking top-level files ... WARNING
A complete check needs the 'checkbashisms' script.
See section ‘Configure and cleanup’ in the ‘Writing R Extensions’
manual.

* checking examples ... NOTE
Examples with CPU (user + system) or elapsed time > 5s
                      user system elapsed
lgb.cv              23.612  0.098  66.864
lgb.model.dt.tree   12.887  0.039  35.043
lgb.importance       7.528  0.051  19.835
lgb.plot.importance  7.190  0.028  20.008
lgb.interprete       5.828  0.031  16.085
lgb.get.eval.result  4.639  0.012  13.082
predict.lgb.Booster  4.547  0.038  12.532
lgb.unloader         4.558  0.017  12.948
lgb.train            4.512  0.049  12.857
```

</details>

⚠️ **`Debian Linux, R-devel, GCC, no long double`**

<details><summary>3 NOTEs, 1 WARNINGS, 0 ERRORs..build timed out</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-e447f0d507334538bb45a9fdab2fdfd3)

```text
* checking installed package size ... NOTE
  installed size is 11.1Mb
  sub-directories of 1Mb or more:
    libs  10.5Mb

* checking top-level files ... WARNING
A complete check needs the 'checkbashisms' script.
See section ‘Configure and cleanup’ in the ‘Writing R Extensions’
manual.

* checking examples ... NOTE
Examples with CPU (user + system) or elapsed time > 5s
                      user system elapsed
lgb.cv              22.500  0.259  64.849
lgb.model.dt.tree   12.275  0.155  43.373
lgb.plot.importance  7.729  0.074  28.011
lgb.importance       7.089  0.105  21.387
lgb.interprete       5.348  0.036  17.705
predict.lgb.Booster  4.418  0.021  17.406
lgb.get.eval.result  4.353  0.032  13.378
lgb.train            4.263  0.105  11.392
lgb.unloader         3.896  0.041  16.054
```

</details>

⚠️ **`Debian Linux, R-devel, clang, ISO-8859-15 locale`**

<details><summary>3 NOTEs, 1 WARNING, 0 ERRORs</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-1187567c8036469d992c1889b582f5f5)

```text
* checking installed package size ... NOTE
  installed size is 11.0Mb
  sub-directories of 1Mb or more:
    libs  10.4Mb

* checking top-level files ... WARNING
A complete check needs the 'checkbashisms' script.
See section 'Configure and cleanup' in the 'Writing R Extensions'
manual.

* checking examples ... NOTE
Examples with CPU (user + system) or elapsed time > 5s
                user system elapsed
lgb.cv         2.106  0.058   6.457
lgb.interprete 1.858  0.037   5.495
```

</details>

❌ **`Debian Linux, R-devel, GCC ASAN/UBSAN`**

<details><summary>failed before getting to R CMD CHECK</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-40375e2cf86e487891de62cbe5261bb2)

many of these:

```
g++ -fsanitize=undefined,bounds-strict -fno-omit-frame-pointer -std=gnu++11 -I"/usr/local/lib/R/include" -DNDEBUG -I./include -DMM_PREFETCH=1 -DMM_MALLOC=1 -DUSE_SOCKET -DLGB_R_BUILD  -I/usr/local/include   -pthread -fpic  -g -O2 -Wall -pedantic -mtune=native -c application/application.cpp -o application/application.o
In file included from ./include/LightGBM/config.h:16:0,
                 from ./include/LightGBM/application.h:8,
                 from application/application.cpp:5:
./include/LightGBM/utils/common.h:775:0: warning: ignoring #pragma omp parallel [-Wunknown-pragmas]
 #pragma omp parallel for schedule(static, 1)
 
./include/LightGBM/utils/common.h:791:0: warning: ignoring #pragma omp parallel [-Wunknown-pragmas]
     #pragma omp parallel for schedule(static, 1)
 
In file included from ./include/LightGBM/dataset.h:9:0,
                 from application/application.cpp:8:
./include/LightGBM/feature_group.h:204:0: warning: ignoring #pragma omp parallel [-Wunknown-pragmas]
       #pragma omp parallel for schedule(guided)
 
In file included from application/application.cpp:8:0:
./include/LightGBM/dataset.h:321:0: warning: ignoring #pragma omp parallel [-Wunknown-pragmas]
 #pragma omp parallel for schedule(static)

application/application.cpp:221:0: warning: ignoring #pragma omp parallel [-Wunknown-pragmas]
     #pragma omp parallel for schedule(static)
 
In file included from ./include/LightGBM/utils/common.h:9:0,
                 from ./include/LightGBM/config.h:16,
                 from ./include/LightGBM/application.h:8,
                 from application/application.cpp:5:
./include/LightGBM/utils/openmp_wrapper.h:84:2: warning: extra ‘;’ [-Wpedantic]
 };  // extern "C"
  ^
g++ -fsanitize=undefined,bounds-strict -fno-omit-frame-pointer -std=gnu++11 -I"/usr/local/lib/R/include" -DNDEBUG -I./include -DMM_PREFETCH=1 -DMM_MALLOC=1 -DUSE_SOCKET -DLGB_R_BUILD  -I/usr/local/include   -pthread -fpic  -g -O2 -Wall -pedantic -mtune=native -c boosting/boosting.cpp -o boosting/boosting.o
In file included from ./include/LightGBM/config.h:16:0,
                 from ./include/LightGBM/boosting.h:8,
                 from boosting/boosting.cpp:5:
./include/LightGBM/utils/common.h:775:0: warning: ignoring #pragma omp parallel [-Wunknown-pragmas]
 #pragma omp parallel for schedule(static, 1)
 
./include/LightGBM/utils/common.h:791:0: warning: ignoring #pragma omp parallel [-Wunknown-pragmas]
     #pragma omp parallel for schedule(static, 1)
```

and  then some errors like this:

```text
[LightGBM] [Info] Saving data to binary file /tmp/RtmpR4zZHQ/lgb.Dataset_b375cc1023c
io/dataset_loader.cpp:340:59: runtime error: reference binding to misaligned address 0x7ff62587302e for type 'const value_type', which requires 4 byte alignment
0x7ff62587302e: note: pointer points here
 00 00 01 00 13 00  00 00 ff ff ff ff 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff
             ^ 
/usr/include/c++/7/bits/stl_vector.h:948:21: runtime error: reference binding to misaligned address 0x7ff62587302e for type 'const int', which requires 4 byte alignment
0x7ff62587302e: note: pointer points here
 00 00 01 00 13 00  00 00 ff ff ff ff 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff
             ^ 
/usr/include/c++/7/bits/vector.tcc:415:28: runtime error: reference binding to misaligned address 0x7ff62587302e for type 'const type', which requires 4 byte alignment
0x7ff62587302e: note: pointer points here
 00 00 01 00 13 00  00 00 ff ff ff ff 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff
             ^ 
/usr/include/c++/7/bits/move.h:74:36: runtime error: reference binding to misaligned address 0x7ff62587302e for type 'const int', which requires 4 byte alignment
0x7ff62587302e: note: pointer points here
 00 00 01 00 13 00  00 00 ff ff ff ff 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff
             ^ 
/usr/include/c++/7/bits/alloc_traits.h:475:4: runtime error: reference binding to misaligned address 0x7ff62587302e for type 'const type', which requires 4 byte alignment
0x7ff62587302e: note: pointer points here
 00 00 01 00 13 00  00 00 ff ff ff ff 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff
             ^ 
/usr/include/c++/7/bits/move.h:74:36: runtime error: reference binding to misaligned address 0x7ff62587302e for type 'const int', which requires 4 byte alignment
0x7ff62587302e: note: pointer points here
 00 00 01 00 13 00  00 00 ff ff ff ff 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff
             ^ 
/usr/include/c++/7/ext/new_allocator.h:136:46: runtime error: reference binding to misaligned address 0x7ff62587302e for type 'const type', which requires 4 byte alignment
0x7ff62587302e: note: pointer points here
 00 00 01 00 13 00  00 00 ff ff ff ff 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff
             ^ 
/usr/include/c++/7/bits/move.h:74:36: runtime error: reference binding to misaligned address 0x7ff62587302e for type 'const int', which requires 4 byte alignment
0x7ff62587302e: note: pointer points here
 00 00 01 00 13 00  00 00 ff ff ff ff 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff
             ^ 
/usr/include/c++/7/ext/new_allocator.h:136:4: runtime error: load of misaligned address 0x7ff62587302e for type 'const int', which requires 4 byte alignment
0x7ff62587302e: note: pointer points here
 00 00 01 00 13 00  00 00 ff ff ff ff 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff
             ^ 
/usr/include/c++/7/bits/stl_vector.h:943:30: runtime error: reference binding to misaligned address 0x7ff62587303a for type 'const int', which requires 4 byte alignment
0x7ff62587303a: note: pointer points here
 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff 15 00  00 00 16 00 00 00
              ^ 
/usr/include/c++/7/bits/alloc_traits.h:475:4: runtime error: reference binding to misaligned address 0x7ff62587303a for type 'const type', which requires 4 byte alignment
0x7ff62587303a: note: pointer points here
 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff 15 00  00 00 16 00 00 00
              ^ 
/usr/include/c++/7/bits/move.h:74:36: runtime error: reference binding to misaligned address 0x7ff62587303a for type 'const int', which requires 4 byte alignment
0x7ff62587303a: note: pointer points here
 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff 15 00  00 00 16 00 00 00
              ^ 
/usr/include/c++/7/ext/new_allocator.h:136:46: runtime error: reference binding to misaligned address 0x7ff62587303a for type 'const type', which requires 4 byte alignment
0x7ff62587303a: note: pointer points here
 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff 15 00  00 00 16 00 00 00
              ^ 
/usr/include/c++/7/bits/move.h:74:36: runtime error: reference binding to misaligned address 0x7ff62587303a for type 'const int', which requires 4 byte alignment
0x7ff62587303a: note: pointer points here
 14 00  00 00 ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff 15 00  00 00 16 00 00 00
```

</details>

❌ **`CentOS 6 with Redhat Developer Toolset, R from EPEL`**

<details><summary>3 NOTEs, 1 WARNING, 1 ERROR</summary>

[logs](https://builder.r-hub.io/status/original/lightgbm_2.3.2.tar.gz-53a62ca38f1145b19418cf25772f228b)

```text
* checking installed package size ... NOTE
  installed size is 36.2Mb
  sub-directories of 1Mb or more:
    libs  35.6Mb

* checking compilation flags used ... WARNING
Compilation used the following non-portable flag(s):
  ‘-Wp,-D_FORTIFY_SOURCE=2’

* checking examples ... NOTE
Examples with CPU or elapsed time > 5s
                      user system elapsed
lgb.cv              17.445  0.110  48.786
lgb.model.dt.tree   12.655  0.039  36.095
lgb.plot.importance  7.178  0.027  20.485
lgb.importance       7.166  0.028  20.537
lgb.interprete       5.120  0.014  14.479
lgb.get.eval.result  4.393  0.012  12.221
lgb.train            4.348  0.016  11.936
predict.lgb.Booster  4.301  0.014  12.238
lgb.unloader         4.206  0.030  12.069

* checking tests ...
  Running ‘testthat.R’ [7m/20m]
 ERROR
Running the tests in ‘tests/testthat.R’ failed.
Last 13 lines of output:
  You can set `force_row_wise=true` to remove the overhead.
  And if memory is not enough, you can set `force_col_wise=true`.
  [LightGBM] [Info] Total Bins 232
  [LightGBM] [Info] Number of data points in the train set: 6513, number of used features: 116
  [LightGBM] [Info] Start training from score 0.482113
  [LightGBM] [Warning] No further splits with positive gain, best gain: -inf
  ══ testthat results  ═══════════════════════════════════════════════════════════
  [ OK: 765 | SKIPPED: 2 | WARNINGS: 0 | FAILED: 4 ]
  1. Failure: lgb.train() works as expected with sparse features (@test_basic.R#477) 
  2. Failure: learning-to-rank with lgb.cv() works as expected (@test_learning_to_rank.R#119) 
  3. Failure: learning-to-rank with lgb.cv() works as expected (@test_learning_to_rank.R#125) 
  4. Failure: learning-to-rank with lgb.cv() works as expected (@test_learning_to_rank.R#131) 
  
  Error: testthat unit tests failed
  Execution halted
```

</details>

## Ignored warnings and notes

The following notes and warnings can be safely ignored and should not block us on CRAN.

```
* checking CRAN incoming feasibility ... NOTE
Maintainer: ‘Guolin Ke <guolin.ke@microsoft.com>’

New submission

License components with restrictions and base license permitting such:
  MIT + file LICENSE
```

```
* checking whether package ‘lightgbm’ can be installed ... WARNING
Found the following significant warnings:
  Warning: package ‘R6’ was built under R version 4.0.2
See ‘/Users/userllqNxUln/lightgbm.Rcheck/00install.out’ for details.
```

```
* checking top-level files ... WARNING
A complete check needs the 'checkbashisms' script.
See section ‘Configure and cleanup’ in the ‘Writing R Extensions’
manual.
```

```
* checking installed package size ... NOTE
  installed size is 35.7Mb
  sub-directories of 1Mb or more:
    libs  35.1Mb
```

--- a/R-package/README.md
+++ b/R-package/README.md
@@ -49,19 +49,15 @@ model <- lgb.cv(
 
 ### Installing the CRAN package
 
-As of this writing, `LightGBM`'s R package is not available on CRAN. However, start with `LightGBM` 3.0.0, you can install a released source distribution. This is the same type of package that you'd install from CRAN. It does not require `CMake`, `Visual Studio`, or anything else outside the CRAN toolchain.
-
-To install this package on any operating system:
-
-1. Choose a release from [the "Releases" page](https://github.com/microsoft/LightGBM/releases).
-2. Look for the artifact with a name like `lightgbm-{VERSION}-r-cran.tar.gz`. Right-click it and choose "copy link address".
-3. Copy that link into `PKG_URL` in the code below and run it.
+`{lightgbm}` is [available on CRAN](https://cran.r-project.org/package=lightgbm), and can be installed with the following R code.
 
 ```r
-PKG_URL <- "https://github.com/microsoft/LightGBM/releases/download/v3.0.0/lightgbm-3.0.0-r-cran.tar.gz"
-
-remotes::install_url(PKG_URL)
+install.packages("lightgbm", repos = "https://cran.r-project.org")
 ```
+
+This is the easiest way to install `{lightgbm}`. It does not require `CMake` or `Visual Studio`, and  should work well on many different operating systems and compilers.
+
+Each CRAN package is also available on [LightGBM releases](https://github.com/microsoft/LightGBM/releases), with a name like `lightgbm-{VERSION}-r-cran.tar.gz`.
 
 #### Custom Installation (Linux, Mac)
 
@@ -322,73 +318,6 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
 1. Edit `configure.win` directly.
 2. Edit `src/Makevars.win.in`.
 
-### Build Precompiled Binaries of the CRAN Package
-
-This section is mainly for maintainers. As long as the R package is not available on CRAN (which will build precompiled binaries automatically) you may want to build precompiled versions of the R package manually, since these will be easier for users to install.
-
-For more details, see ["Writing R Extensions"](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Building-binary-packages).
-
-Packages built like this will only work for the minor version of R used to build them. They may or may not work across different versions of operating systems.
-
-**Mac**
-
-Binary produced: `lightgbm-${VERSION}-r40-macos.tgz`.
-
-```shell
-LGB_VERSION="3.0.0-1"
-sh build-cran-package.sh
-R CMD INSTALL --build lightgbm_${LGB_VERSION}.tar.gz
-mv \
-    lightgbm_${LGB_VERSION}.tgz \
-    lightgbm-${LGB_VERSION}-r40-macos.tgz
-```
-
-**Linux**
-
-Binary produced: `lightgbm-${VERSION}-r40-linux.tgz`.
-
-You can access a Linux system that has R and its build toolchain installed with the `rocker` Docker images.
-
-```shell
-R_VERSION=4.0.2
-
-docker run \
-    -v $(pwd):/opt/LightGBM \
-    -it rocker/verse:${R_VERSION} \
-        /bin/bash
-```
-
-From inside that container, the commands to create a precompiled binary are very similar.
-
-```shell
-cd /opt/LightGBM
-LGB_VERSION="3.0.0-1"
-sh build-cran-package.sh
-R CMD INSTALL --build lightgbm_${LGB_VERSION}.tar.gz
-mv \
-    lightgbm_${LGB_VERSION}_R_*-linux-gnu.tar.gz \
-    lightgbm-${LGB_VERSION}-r40-linux.tgz
-```
-
-Exit the container, and the binary package should still be there on the host system.
-
-```shell
-exit
-```
-
-**Windows**
-
-Binary produced: `lightgbm-${VERSION}-r40-windows.zip`.
-
-```shell
-LGB_VERSION="3.0.0-1"
-sh build-cran-package.sh
-R CMD INSTALL --build lightgbm_${LGB_VERSION}.tar.gz
-mv \
-    lightgbm_${LGB_VERSION}.tgz \
-    lightgbm-${LGB_VERSION}-r40-windows.zip
-```
-
 ### Testing the CRAN Package
 
 `{lightgbm}` is tested automatically on every commit, across many combinations of operating system, R version, and compiler. This section describes how to test the package locally while you are developing.
@@ -495,3 +424,12 @@ Known Issues
 ------------
 
 For information about known issues with the R package, see the [R-package section of LightGBM's main FAQ page](https://lightgbm.readthedocs.io/en/latest/FAQ.html#r-package).
+
+Outdated Links
+--------------
+
+This section is used to redirect people following links that have been removed from the documentation above.
+
+### Build Precompiled Binaries of the CRAN Package
+
+Prior to `{lightgbm}` making it to CRAN, precompiled binaries for the R package were built manually and added to releases. CRAN does this automatically, so this section was removed.

--- a/R-package/README.md
+++ b/R-package/README.md
@@ -405,12 +405,3 @@ Known Issues
 ------------
 
 For information about known issues with the R package, see the [R-package section of LightGBM's main FAQ page](https://lightgbm.readthedocs.io/en/latest/FAQ.html#r-package).
-
-Outdated Links
---------------
-
-This section is used to redirect people following links that have been removed from the documentation above.
-
-### Build Precompiled Binaries of the CRAN Package
-
-Prior to `{lightgbm}` making it to CRAN, precompiled binaries for the R package were built manually and added to releases. CRAN does this automatically, so this section was removed.

--- a/R-package/README.md
+++ b/R-package/README.md
@@ -171,38 +171,19 @@ Rscript build_r.R --use-gpu
 
 ### Installing Precompiled Binaries
 
-**NOTE:** As of this writing, the precompiled binaries of the R package should be considered experimental. If you try them an experience any problems, please [open an issue](https://github.com/microsoft/LightGBM/issues).
-
-Starting with `LightGBM` 3.0.0, precompiled binaries for the R package are created for each release. These packages do not require compilation, so they will be faster and easier to install than packages that are built from source. These packages are created with R 4.0 and are not guaranteed to work with other R versions.
-
-Binaries are available for Windows, Mac, and Linux systems. They are not guaranteed to work with all variants and versions of these operating systems. Please [open an issue](https://github.com/microsoft/LightGBM/issues) if you encounter any problems.
-
-To install a binary for the R package:
-
-1. Choose a release from [the "Releases" page](https://github.com/microsoft/LightGBM/releases).
-2. Choose a file based on your operating system. Right-click it and choose "copy link address".
-    * Linux: `lightgbm-{VERSION}-r40-linux.tgz`
-    * Mac: `lightgbm-{VERSION}-r40-macos.tgz`
-    * Windows: `lightgbm-{VERSION}-r40-windows.zip`
-3. Copy that link into `PKG_URL` in the code below and run it.
-
-This sample code installs version 3.0.0-1 of the R package on Mac.
+Precompiled binaries for Mac and Windows are prepared by CRAN a few days after each release to CRAN. They can be installed with the following R code.
 
 ```r
-PKG_URL <- "https://github.com/microsoft/LightGBM/releases/download/v3.0.0rc1/lightgbm-3.0.0-1-r40-macos.tgz"
-
-local_file <- paste0("lightgbm.", tools::file_ext(PKG_URL))
-
-download.file(
-    url = PKG_URL
-    , destfile = local_file
-)
 install.packages(
-    pkgs = local_file
-    , type = "binary"
-    , repos = NULL
+    "lightgbm"
+    , type = "both"
+    , repos = "https://cran.r-project.org"
 )
 ```
+
+These packages do not require compilation, so they will be faster and easier to install than packages that are built from source.
+
+CRAN does not prepare precompiled binaries for Linux, and as of this writing neither does this project.
 
 ### Installing from a Pre-compiled lib_lightgbm <a name="lib_lightgbm"></a>
 

